From 96eb881f891799a98a72ea948458a5d696b3adce Mon Sep 17 00:00:00 2001
From: Misael Lopez Cruz <misael.lopez@ti.com>
Date: Mon, 7 Feb 2011 23:12:45 -0600
Subject: mfd: twl6040: Add set_bits and clear_bits functions

Add two new APIs to twl6040 MFD:
- twl6040_set_bits   : Set bits specified in a mask
- twl6040_clear_bits : Clear bits specified in a mask

Change-Id: Ib5b7ae54469b8030e67f25bbf309fa70dfdefe27
Signed-off-by: Misael Lopez Cruz <misael.lopez@ti.com>
---
 drivers/mfd/twl6040-codec.c       |   36 ++++++++++++++++++++++++++++++++++++
 include/linux/mfd/twl6040-codec.h |    4 ++++
 2 files changed, 40 insertions(+), 0 deletions(-)

diff --git a/drivers/mfd/twl6040-codec.c b/drivers/mfd/twl6040-codec.c
index 193fd7f..9afb051 100644
--- a/drivers/mfd/twl6040-codec.c
+++ b/drivers/mfd/twl6040-codec.c
@@ -63,6 +63,42 @@ int twl6040_reg_write(struct twl6040_codec *twl6040, unsigned int reg, u8 val)
 }
 EXPORT_SYMBOL(twl6040_reg_write);
 
+int twl6040_set_bits(struct twl6040_codec *twl6040, unsigned int reg, u8 mask)
+{
+	int ret;
+	u8 val;
+
+	mutex_lock(&twl6040->io_mutex);
+	ret = twl_i2c_read_u8(TWL_MODULE_AUDIO_VOICE, &val, reg);
+	if (ret)
+		goto out;
+
+	val |= mask;
+	ret = twl_i2c_write_u8(TWL_MODULE_AUDIO_VOICE, val, reg);
+out:
+	mutex_unlock(&twl6040->io_mutex);
+	return ret;
+}
+EXPORT_SYMBOL(twl6040_set_bits);
+
+int twl6040_clear_bits(struct twl6040_codec *twl6040, unsigned int reg, u8 mask)
+{
+	int ret;
+	u8 val;
+
+	mutex_lock(&twl6040->io_mutex);
+	ret = twl_i2c_read_u8(TWL_MODULE_AUDIO_VOICE, &val, reg);
+	if (ret)
+		goto out;
+
+	val &= ~mask;
+	ret = twl_i2c_write_u8(TWL_MODULE_AUDIO_VOICE, val, reg);
+out:
+	mutex_unlock(&twl6040->io_mutex);
+	return ret;
+}
+EXPORT_SYMBOL(twl6040_clear_bits);
+
 /* twl6040 codec manual power-up sequence */
 static int twl6040_power_up(struct twl6040_codec *twl6040)
 {
diff --git a/include/linux/mfd/twl6040-codec.h b/include/linux/mfd/twl6040-codec.h
index 8840f9c..b976b8a 100644
--- a/include/linux/mfd/twl6040-codec.h
+++ b/include/linux/mfd/twl6040-codec.h
@@ -224,6 +224,10 @@ static inline void twl6040_free_irq(struct twl6040_codec *twl6040, int irq,
 int twl6040_reg_read(struct twl6040_codec *twl6040, unsigned int reg);
 int twl6040_reg_write(struct twl6040_codec *twl6040, unsigned int reg,
 		      u8 val);
+int twl6040_set_bits(struct twl6040_codec *twl6040, unsigned int reg,
+		     u8 mask);
+int twl6040_clear_bits(struct twl6040_codec *twl6040, unsigned int reg,
+		       u8 mask);
 int twl6040_enable(struct twl6040_codec *twl6040);
 int twl6040_disable(struct twl6040_codec *twl6040);
 int twl6040_is_enabled(struct twl6040_codec *twl6040);
-- 
1.6.0.4

